### PR TITLE
[AI] tweak research of hull lines

### PR DESCRIPTION
**Pull request by MichaelCourtney**
W.I.P.
I've tested that it loads and queues the techs in the logs, but haven't play tested further.
AIs below typical aggression/AIs in "sparse galaxy" are identical to before commit
AIs above typical aggression are identical to before commit except for hulls (and hull pre-reqs)
Number of hulls in queue should be lower (after initial organic/robotic) and follow only 1 - 2 hull lines
Aim is to be a starting point with least changes, with further tweaks to follow

### CODE REVIEW

**Comment by MichaelCourtney** [comment]
hmm judging by the linter test failing, this note in test_tech_integrity:
"Any error is displayed in chat window."
is simply not true, or out of date.

**Comment by MichaelCourtney** [comment]
"Galaxy is Sparse" only triggers on no monsters, large galaxy, few players. I'd like to look at removing those lists

**Comment by MichaelCourtney** [comment]
Watched the AI for about 75 turns or so. They use their fancy heavy hulls to spam chaff troopers, but I reverted the changes and watched them again, they do that anyway. I'll look into it.

**Comment by MichaelCourtney** [comment]
Had a sudden hunch and I tried again with specials/natives off and suddenly they were building warships. They get to a certain point and then start trooper spamming again (maybe they decide they have "enough" warship).

I have a suspicion the ancient guardians are triggering them.

**Comment by Cjkjvfnby** [comment]
> hmm judging by the linter test failing, this note in test_tech_integrity:
   "Any error is displayed in chat window."
   is simply not true, or out of date.

I have fixed this.  But because of message order, the message is above the greetings, so it's easy to miss it.  

What is your opinion should it be fixed or could we remove outdated documentation? https://github.com/freeorion/freeorion/pull/4648

**Comment by MichaelCourtney** [comment]
A small note that current AI doesn't build the final asteroid shipyard, on the grounds the ship designer values titans more

**Comment by MichaelCourtney** [comment]
Looking into invasion, having a cheaper hull decreases the cost of troopers which in turn increases the chance of an invasion target passing the checks, so possibly some extra trooper spamming

File: default/python/AI/ResearchAI.py [hunk 1]
```diff
@@ -145,6 +145,7 @@ def generate_classic_research_orders():  # noqa: max-complexity
     aistate = get_aistate()
     enemies_sighted = aistate.misc.get("enemies_sighted", {})
     galaxy_is_sparse = ColonisationAI.galaxy_is_sparse()
+    galaxy_is_cramped = ColonisationAI.galaxy_is_cramped()
```
Comment (Cjkjvfnby) [inline default/python/AI/ResearchAI.py:148]: You could inline this variable.

File: default/python/AI/character/character_module.py [hunk 2]
```diff
@@ -543,11 +543,11 @@ def may_travel_beyond_supply(self, distance):
     # TODO remove this function as soon as old style research is gone
     # It defeats the orthogonality goal of character components for little gain
     def get_research_index(self):
-        research_index = self.id % 2
+        research_index = 0
         if self.aggression >= fo.aggression.aggressive:
             research_index = 2 + (self.id % 3)  # so indices [2,3,4]
         elif self.aggression >= fo.aggression.typical:
-            research_index += 1
+            research_index = 1 + (self.id % 2)  # indices [1,2,3]
```
Comment (Grummel7) [inline default/python/AI/character/character_module.py:550]: Using %2, this gives 1 or 2, never 3.
I never looked at this function before, but seeing it no, I wonder what it should be good for.
Should we really maintain so many research tables? I think there are easier ways to make them less aggressive.

**Comment by Grummel7** [comment]
> Watched the AI for about 75 turns or so. They use their fancy heavy hulls to spam chaff troopers, but I reverted the changes and watched them again, they do that anyway. I'll look into it.

The ship design selection definitely needs work, too.

I tested this PR on top of master for about 300 turns twice. In the first game, although the AI had titans and heavy asteroid hulls, battle ships used nothing bigger than plain asteroid hulls. Only one AI built a Geo-Integration facility, but it was blockaded by a stronger AI, when I checked it. In the second game, at least one AI used heavy asteroid hulls.

First question is why they did not use heavy asteroid hull, which can be built, wherever asteroid hulls can be built.

Second question is, why they did not build a Geo-Int. => #4778 should fix it.

I'll investigate into the second question.

**Comment by MichaelCourtney** [comment]
Decided to breakup reworking the tech tree into smaller tweaks

**Comment by Vezzra** [comment]
> Decided to breakup reworking the tech tree into smaller tweaks

Meaning, this PR is going to be superseded by a couple of smaller ones, correct? Can you post here which PRs that are?

**Comment by MichaelCourtney** [comment]
> > Decided to breakup reworking the tech tree into smaller tweaks
> 
> Meaning, this PR is going to be superseded by a couple of smaller ones, correct? Can you post here which PRs that are?

1st is here:

https://github.com/freeorion/freeorion/pull/4879

2nd doesn't exist yet, as the plan is to do it with small incremental tweaks

**Pull request closed without merging**